In [1]:
!pip install pandas -q


#Importation des bibliothèques nécessaires
import pandas as pd
from pathlib import Path        


In [2]:
#Ouvertude des premires datasets

In [3]:
#Ouvertude / chargements des premires datasets


CountrySeries1= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsCountry-Series.csv")
StatsCountry2= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsCountry.csv")
StatsData3= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsData.csv")
StatsFootNote4= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsFootNote.csv")
StatsSeries5= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsSeries.csv")

CountrySeries1= pd.read_csv(CountrySeries1)
StatsCountry2= pd.read_csv(StatsCountry2)
StatsData3= pd.read_csv(StatsData3)
StatsFootNote4= pd.read_csv(StatsFootNote4)
StatsSeries5= pd.read_csv(StatsSeries5)


In [4]:
x == 6

NameError: name 'x' is not defined

## Première partie : Analyse initiale des datasets – World Bank EdStats


---

### Détails des datasets

### 1. CountrySeries1  
**Shape : (613, 4)**  

Colonnes :  
- CountryCode  
- SeriesCode  
- DESCRIPTION  
- Unnamed  

Il s'agit d'une table de correspondance entre pays et indicateur, avec une description associée.

Cette table permet d’identifier, pour un pays donné et un indicateur donné, la source ou la spécificité méthodologique utilisée.

### Utilité :
- Vérifier la qualité des sources par pays  
- Détecter des différences méthodologiques  
- Justifier d’éventuels écarts entre pays  

---

### 2. StatsCountry2  
**Shape : (241, 32)**  

Il s'agit de la Dimension Pays.

Chaque ligne correspond à un pays, une région ou un groupe de pays.

Ce dataset contient des informations générales sur les pays :

- Données économiques  
- Groupe de revenu (Income Group)  
- Région  
- Monnaie  
- Standards IMF  
- Données sur les derniers recensements  
- Informations agricoles, industrielles, commerciales  
- Données administratives et statistiques  

À noter :  
Il y a 241 régions/pays/groupes de pays selon la classification habituelle, incluant des agrégats comme Arab World, Sub-Saharan Africa, etc.

#### Utilité :
- Analyse fine par groupe de pays  
- Segmentation par niveau de revenu  
- Construction de clusters  
- Croisement avec les données de population et d’éducation  

C’est un dataset clé pour l’analyse stratégique.

---

### 3. StatsData3  
**Shape : (886930, 70)**  

Le troisième dataset est le plus volumineux.

Il contient les données principales relatives :

- À la population (par tranche d’âge)  
- Au niveau d’éducation (par diplôme, genre, etc.)  
- Aux dépenses  
- À d’autres indicateurs éducatifs  

#### Structure :

Chaque ligne correspond à :  
1 pays × 1 indicateur  

On observe :  
241 pays × 3665 indicateurs ≈ 886 930 observations  

Les colonnes années vont de 1970 à 2100.

- 1970–2017 : données historiques  
- 2020–2100 : projections  
- Les années 2018–2019 sont absentes (probablement lié à la structure du dataset)

Cela représente une série temporelle de :

- 46 années historiques  
- 18 années projetées  
- Soit 64 points temporels  

Chaque colonne correspond à une année et contient la valeur de l’indicateur.

#### Grandes familles d’indicateurs :

1. Education  
2. Population  
3. Dépenses  

Ce dataset constitue le cœur de la matrice des données pour notre étude.

Il est relié aux autres dimensions :

- Pays → StatsCountry2  
- Indicateurs → StatsSeries5  
- Notes → StatsFootNote4  

Il présente un niveau de granularité extrêmement fin.

---

### 4. StatsFootNote4  
**Shape : (643637, 5)**  

Ce dataset contient les métadonnées spécifiques par observation.

Il s’agit des notes de bas de page associées à :  
1 pays + 1 indicateur + 1 année  

On observe environ 239 codes pays distincts.

Ce fichier apporte un complément sur :

- Les sources  
- Les estimations  
- Les précisions méthodologiques  
- Les particularités nationales  

Il est important pour qualifier la fiabilité des données.

---

### 5. StatsSeries5  
**Shape : (3665, 21)**  

Il correspond à la Dimension Indicateur.

Chaque ligne représente un indicateur unique.

Ce dataset permet de comprendre :

- Ce que mesure réellement l’indicateur  
- Sa définition courte et longue  
- Son thème (Topic)  
- Sa source  
- Sa méthodologie  
- Son unité de mesure  

Il est essentiel pour interpréter correctement les valeurs présentes dans StatsData3.

---

## Synthèse générale


Nous disposons :

- De 241 pays / régions  
- De 3665 indicateurs uniques  
- De données couvrant la période 1970–2100  
- D’un total de 886 930 combinaisons pays × indicateurs  

La structure globale correspond à un modèle en étoile (star schema) :

StatsSeries5 (Dimension Indicateur)  
        ↑  
StatsCountry2 ← StatsData3 → StatsFootNote4  
        ↓  
CountrySeries1  

StatsData3 est la table centrale autour de laquelle s’articulent les dimensions pays, indicateurs et notes.


In [ ]:

#Récupération des noms de colonnes et des dimensions de chaque DataFrame

names = [
    "CountrySeries1","StatsCountry2", "StatsData3", "StatsFootNote4", "StatsSeries5",
]

for name in names:
    df = globals().get(name)
    if isinstance(df, pd.DataFrame):
        print(f"{name} — shape: {df.shape}")
        print(df.columns.tolist())
        print()
    else:
        print(f"{name} introuvable ou n'est pas un DataFrame.\n")

CountrySeries1 — shape: (613, 4)
['CountryCode', 'SeriesCode', 'DESCRIPTION', 'Unnamed: 3']

StatsCountry2 — shape: (241, 32)
['Country Code', 'Short Name', 'Table Name', 'Long Name', '2-alpha code', 'Currency Unit', 'Special Notes', 'Region', 'Income Group', 'WB-2 code', 'National accounts base year', 'National accounts reference year', 'SNA price valuation', 'Lending category', 'Other groups', 'System of National Accounts', 'Alternative conversion factor', 'PPP survey year', 'Balance of Payments Manual in use', 'External debt Reporting status', 'System of trade', 'Government Accounting concept', 'IMF data dissemination standard', 'Latest population census', 'Latest household survey', 'Source of most recent Income and expenditure data', 'Vital registration complete', 'Latest agricultural census', 'Latest industrial data', 'Latest trade data', 'Latest water withdrawal data', 'Unnamed: 31']

StatsData3 — shape: (886930, 70)
['Country Name', 'Country Code', 'Indicator Name', 'Indicator C

In [ ]:
# Faire apparaître les deux premières lignes de chaque DataFrame

datasets = {
    'CountrySeries1': CountrySeries1,
    'StatsCountry2': StatsCountry2,
    'StatsData3': StatsData3,
    'StatsFootNote4': StatsFootNote4,
    'StatsSeries5': StatsSeries5
}

for dataset_name, df in datasets.items():
    print(f"\n{'='*80}")
    print(f"DATASET: {dataset_name}")
    print(f"{'='*80}\n")
    
    for row_idx in range(min(2, len(df))):
        print(f"\n--- Row {row_idx} ---")
        for col_name in df.columns:
            cell_value = df.loc[row_idx, col_name]
            print(f"{col_name}: {cell_value}")


DATASET: CountrySeries1


--- Row 0 ---
CountryCode: ABW
SeriesCode: SP.POP.TOTL
DESCRIPTION: Data sources : United Nations World Population Prospects
Unnamed: 3: nan

--- Row 1 ---
CountryCode: ABW
SeriesCode: SP.POP.GROW
DESCRIPTION: Data sources: United Nations World Population Prospects
Unnamed: 3: nan

DATASET: StatsCountry2


--- Row 0 ---
Country Code: ABW
Short Name: Aruba
Table Name: Aruba
Long Name: Aruba
2-alpha code: AW
Currency Unit: Aruban florin
Special Notes: SNA data for 2000-2011 are updated from official government statistics; 1994-1999 from UN databases. Base year has changed from 1995 to 2000.
Region: Latin America & Caribbean
Income Group: High income: nonOECD
WB-2 code: AW
National accounts base year: 2000
National accounts reference year: nan
SNA price valuation: Value added at basic prices (VAB)
Lending category: nan
Other groups: nan
System of National Accounts: Country uses the 1993 System of National Accounts methodology.
Alternative conversion factor: nan


##Schéma de l'architecture globale des données (sortie d'un LLM)


![Schéma](image%20schéma/mon_image.png)
![Schéma](image%20schéma/mon_image.png)


### Deuxième partie : début du nettoyage

Partie du code qui correspond à l'étape 2 du cours :
- calcul du nombre de doublons dans le jeu de données, suppression si il y en a
- Calcule de la proportion de valeurs manquantes par colonne et par jeux de données afin de supprimer les valeurs inutilisable
- Determiner la forme de chaque séries : valeur numérique (continue), ou valeur catégorielle
- Determiner les définition de chaque colonnes de chaque datasets

- Traitement de chaque fichier l'un après l'autre

#### Dataset 1 : CountrySeries

In [ ]:
# nombre de valeurs manquantes par colonne dans CountrySeries1 en pourcentage
CountrySeries1.isna().sum() / len(CountrySeries1) * 100

CountryCode      0.0
SeriesCode       0.0
DESCRIPTION      0.0
Unnamed: 3     100.0
dtype: float64

In [ ]:
#Connaître le type de données de chaque colonne dans CountrySeries1
CountrySeries1.dtypes

CountryCode        str
SeriesCode         str
DESCRIPTION        str
Unnamed: 3     float64
dtype: object

In [ ]:
CountrySeries1.duplicated().sum() #Connaitre le nombre de doublons

#Aucune ligne en doublon

np.int64(0)

##### Résumé CountrySeries1 (613, 4)

attention codeMarkdown généré par LLM; non automatique
| Colonne      | Description | % de NA | Type de données |
|-------------|------------|---------|----------------|
| CountryCode | Code ISO 3 lettres du pays (clé de jointure). | 0.0 % | Texte (string / object) |
| SeriesCode  | Code unique de l’indicateur. | 0.0 % | Texte (string / object) |
| DESCRIPTION | Description spécifique liée au pays et à l’indicateur (source ou particularité méthodologique). | 0.0 % | Texte (string / object) |


In [ ]:

CountrySeries1 = CountrySeries1.drop(columns=['Unnamed: 3']) #supprimer la colonne "Unnamed: 3" qui ne contient que des valeurs manquantes


#Calculs automatiquement vérifié avec le viewer de DataFrame de VSCode 

#### Dataset 2: StatsCountry2


Mêmes opérations avec le deuxième datasets


In [ ]:
StatsCountry2.isna().sum() / len(StatsCountry2) * 100

Country Code                                           0.000000
Short Name                                             0.000000
Table Name                                             0.000000
Long Name                                              0.000000
2-alpha code                                           1.244813
Currency Unit                                         10.788382
Special Notes                                         39.834025
Region                                                11.203320
Income Group                                          11.203320
WB-2 code                                              0.414938
National accounts base year                           14.937759
National accounts reference year                      86.721992
SNA price valuation                                   18.257261
Lending category                                      40.248963
Other groups                                          75.933610
System of National Accounts             

In [ ]:
StatsCountry2.duplicated().sum()
#Aucune ligne en doublon

np.int64(0)

In [ ]:
##Data set 2


StatsCountry2 = StatsCountry2.drop(columns=['WB-2 code'])


#### StatsCountry2 (241, 32)
ce tableau donne une description plus claire de chaque colonne. descriptions récupérées autre part

| Colonne | Description | % de NA |
|----------|------------|---------|
| Country Code | Code ISO à 3 lettres du pays (clé de jointure principale). | 0.00 % |
| Short Name | Nom court du pays. | 0.00 % |
| Table Name | Nom utilisé dans les tableaux officiels de la base de données. | 0.00 % |
| Long Name | Nom officiel complet du pays. | 0.00 % |
| 2-alpha code | Code ISO à 2 lettres du pays. | 1.24 % |
| Currency Unit | Devise officielle du pays. | 10.79 % |
| Region | Région géographique (ex : Europe & Central Asia). | 11.20 % |
| Income Group | Groupe de revenu selon la classification Banque mondiale. | 11.20 % |
| National accounts base year | Année de base utilisée pour les comptes nationaux. | 14.94 % |
| National accounts reference year | Année de référence des comptes nationaux. | 86.72 % |
| SNA price valuation | Méthode d’évaluation des prix selon le Système de Comptabilité Nationale (SNA). | 18.26 % |
| Lending category | Catégorie de prêt attribuée par la Banque mondiale. | 40.25 % |
| Other groups | Autres regroupements économiques ou administratifs. | 75.93 % |
| System of National Accounts | Version du système de comptabilité nationale utilisée. | 10.79 % |
| Alternative conversion factor | Facteur alternatif de conversion monétaire. | 80.50 % |
| PPP survey year | Année de l’enquête utilisée pour le calcul des PPA. | 39.83 % |
| Balance of Payments Manual in use | Version du manuel de balance des paiements utilisée. | 24.90 % |
| External debt Reporting status | Statut de déclaration de la dette extérieure. | 48.55 % |
| System of trade | Système de comptabilisation du commerce (général/spécial). | 17.01 % |
| Government Accounting concept | Concept comptable utilisé par le gouvernement. | 33.20 % |
| IMF data dissemination standard | Standard de diffusion des données du FMI respecté. | 24.90 % |
| Latest population census | Année du dernier recensement de la population. | 11.62 % |
| Latest household survey | Année de la dernière enquête auprès des ménages. | 41.49 % |
| Source of most recent Income and expenditure data | Source des dernières données revenus/dépenses. | 33.61 % |
| Vital registration complete | Indique si l’enregistrement des naissances/décès est complet. | 53.94 % |
| Latest agricultural census | Année du dernier recensement agricole. | 41.08 % |
| Latest industrial data | Année des dernières données industrielles disponibles. | 55.60 % |
| Latest trade data | Année des dernières données commerciales disponibles. | 23.24 % |
| Latest water withdrawal data | Année des dernières données sur les prélèvements d’eau. | 25.73 % |
| Unnamed: 31 | Colonne vide issue de l’export (à supprimer). | 100.00 % |


#### Dataset 3

In [ ]:
pd.set_option('display.max_rows', None)
StatsData3.isna().sum()/ len(StatsData3) * 100

#Le nombre de NA dans les colonnes de StatsData3 est très élevé, en particulier dans les colonnes "FootnoteCodes" et "ValueFootnoteSymbol", ce qui peut poser des problèmes pour l'analyse des données. Il serait important de comprendre pourquoi


Country Name        0.000000
Country Code        0.000000
Indicator Name      0.000000
Indicator Code      0.000000
1970               91.849639
1971               95.993258
1972               95.984012
1973               95.992356
1974               95.971497
1975               90.156382
1976               95.773849
1977               95.763589
1978               95.763364
1979               95.849842
1980               89.951631
1981               95.627953
1982               95.770692
1983               95.663694
1984               95.647233
1985               89.819264
1986               95.560867
1987               95.643286
1988               95.653321
1989               95.767422
1990               85.973527
1991               91.607342
1992               91.482642
1993               91.454455
1994               91.266278
1995               85.189248
1996               91.340128
1997               91.718287
1998               90.426076
1999               86.601085
2000          

| Colonne | % Valeurs manquantes |
|----------|----------------------|
| Country Name | 0.00 % |
| Country Code | 0.00 % |
| Indicator Name | 0.00 % |
| Indicator Code | 0.00 % |
| 1970 | 91.85 % |
| 1971 | 95.99 % |
| 1972 | 95.98 % |
| 1973 | 95.99 % |
| 1974 | 95.97 % |
| 1975 | 90.16 % |
| 1976 | 95.77 % |
| 1977 | 95.76 % |
| 1978 | 95.76 % |
| 1979 | 95.85 % |
| 1980 | 89.95 % |
| 1981 | 95.63 % |
| 1982 | 95.77 % |
| 1983 | 95.66 % |
| 1984 | 95.65 % |
| 1985 | 89.82 % |
| 1986 | 95.56 % |
| 1987 | 95.64 % |
| 1988 | 95.65 % |
| 1989 | 95.77 % |
| 1990 | 85.97 % |
| 1991 | 91.61 % |
| 1992 | 91.48 % |
| 1993 | 91.45 % |
| 1994 | 91.27 % |
| 1995 | 85.19 % |
| 1996 | 91.34 % |
| 1997 | 91.72 % |
| 1998 | 90.43 % |
| 1999 | 86.60 % |
| 2000 | 80.08 % |
| 2001 | 86.07 % |
| 2002 | 86.00 % |
| 2003 | 85.30 % |
| 2004 | 85.48 % |
| 2005 | 79.24 % |
| 2006 | 84.18 % |
| 2007 | 84.52 % |
| 2008 | 84.85 % |
| 2009 | 83.98 % |
| 2010 | 72.67 % |
| 2011 | 83.54 % |
| 2012 | 83.40 % |
| 2013 | 84.50 % |
| 2014 | 87.17 % |
| 2015 | 85.22 % |
| 2016 | 98.14 % |
| 2017 | 99.98 % |
| 2020 | 94.20 % |
| 2025 | 94.20 % |
| 2030 | 94.20 % |
| 2035 | 94.20 % |
| 2040 | 94.20 % |
| 2045 | 94.20 % |
| 2050 | 94.20 % |
| 2055 | 94.20 % |
| 2060 | 94.20 % |
| 2065 | 94.20 % |
| 2070 | 94.20 % |
| 2075 | 94.20 % |
| 2080 | 94.20 % |
| 2085 | 94.20 % |
| 2090 | 94.20 % |
| 2095 | 94.20 % |
| 2100 | 94.20 % |
| Unnamed: 69 | 100.00 % |



In [ ]:
StatsData3.duplicated().sum()
#Aucune ligne en doublon

In [ ]:
StatsData3 = StatsData3.drop(columns=['Unnamed: 69'])

#### Dataset 4

In [ ]:
StatsFootNote4.isna().sum() / len(StatsFootNote4) * 100



CountryCode      0.0
SeriesCode       0.0
Year             0.0
DESCRIPTION      0.0
Unnamed: 4     100.0
dtype: float64

| Colonne      | % Valeurs manquantes |
|--------------|----------------------|
| CountryCode  | 0.0 % |
| SeriesCode   | 0.0 % |
| Year         | 0.0 % |
| DESCRIPTION  | 0.0 % |
| Unnamed: 4   | 100.0 % |


In [ ]:
print("Hello, World!")

In [ ]:
StatsFootNote4.dtypes

CountryCode        str
SeriesCode         str
Year               str
DESCRIPTION        str
Unnamed: 4     float64
dtype: object

In [ ]:
StatsFootNote4.duplicated().sum()

#Aucune ligne en doublon

np.int64(0)

In [ ]:
StatsFootNote4 = StatsFootNote4.drop(columns=['Unnamed: 4']) #Supprimer la colonne "Unnamed: 4" qui ne contient que des valeurs manquantes

#### Dataset 5

In [ ]:
StatsSeries5.isna().sum() / len(StatsSeries5) * 100
#Beaucoup de valeurs manquantes dans les colonnes "Short definition" et "Long definition", ce qui peut poser des problèmes pour l'analyse des données. Il serait important de comprendre pourquoi

Series Code                              0.000000
Topic                                    0.000000
Indicator Name                           0.000000
Short definition                        41.173261
Long definition                          0.000000
Unit of measure                        100.000000
Periodicity                             97.298772
Base Period                             91.432469
Other notes                             84.938608
Aggregation method                      98.717599
Limitations and exceptions              99.618008
Notes from original source             100.000000
General comments                        99.618008
Source                                   0.000000
Statistical concept and methodology     99.372442
Development relevance                   99.918145
Related source links                    94.133697
Other web links                        100.000000
Related indicators                     100.000000
License Type                           100.000000



Cinqième dataset

| Colonne                                   | % Valeurs manquantes |
|--------------------------------------------|----------------------|
| Series Code                               | 0.00 % |
| Topic                                     | 0.00 % |
| Indicator Name                            | 0.00 % |
| Short definition                          | 41.17 % |
| Long definition                           | 0.00 % |
| Unit of measure                           | 100.00 % |
| Periodicity                               | 97.30 % |
| Base Period                               | 91.43 % |
| Other notes                               | 84.94 % |
| Aggregation method                        | 98.72 % |
| Limitations and exceptions                | 99.62 % |
| Notes from original source                | 100.00 % |
| General comments                          | 99.62 % |
| Source                                    | 0.00 % |
| Statistical concept and methodology       | 99.37 % |
| Development relevance                     | 99.92 % |
| Related source links                      | 94.13 % |
| Other web links                           | 100.00 % |
| Related indicators                        | 100.00 % |
| License Type                              | 100.00 % |
| Unnamed: 20                               | 100.00 % |


In [ ]:
StatsSeries5.dtypes


Series Code                                str
Topic                                      str
Indicator Name                             str
Short definition                           str
Long definition                            str
Unit of measure                        float64
Periodicity                                str
Base Period                                str
Other notes                                str
Aggregation method                         str
Limitations and exceptions                 str
Notes from original source             float64
General comments                           str
Source                                     str
Statistical concept and methodology        str
Development relevance                      str
Related source links                       str
Other web links                        float64
Related indicators                     float64
License Type                           float64
Unnamed: 20                            float64
dtype: object

In [ ]:
StatsSeries5.duplicated().sum()

np.int64(0)

In [ ]:
StatsSeries5 = StatsSeries5.drop(
    columns=[
        'Unit of measure',
        'Notes from original source',
        'Other web links',
        'Related indicators',
        'License Type',
        'Unnamed: 20'
    ]
) #Supprimer les colonnes 100% vides

Après avoir fait une revue de tous les datasets, on va passer à l'étape 3 les premiers nettoyage

### Etape 3 : Nettoyage 

**résultat attendu** : Code permettant de filtrer les faux pays des tables où cela fait sens
Nous voyons qu'une colonne rescence les méthodes de calcul de chaque pays, sauf pour les groupes de pays où ceux ci sont en NA. Sachant qu'il y a 241 groupes ou pays différents, et que il y a 26 NA dans cette colonne, le dataset comprend donc 241 - 26 = 215, soit un nombre proche du nombre de pays dans le monde


Nous allons filtrer ces observations particulière dans une nouvelle colonne "Groupe Pays" où 1 signie oui, et 0 signifie non.

Cette technique après vérification visuelle  donne un pays malgré tout le pays Nauru qui a un NA dans son système de comptes nationaux défini


In [ ]:
StatsCountry2["Groupe de pays"] = (
    StatsCountry2["Currency Unit"]
    .isna()
    .astype(int)
)


In [ ]:
StatsCountry2[StatsCountry2["Groupe de pays"] == 1].head(30) #Visualisation des 30 premières lignes du groupe de pays 1, qui correspond aux pays n'ayant pas de système de comptes nationaux défini., hors 26 groupe de pays sont à dénombrr



,Country Code,Short Name,Table Name,Long Name,2-alpha code,Currency Unit,Special Notes,Region,Income Group,WB-2 code,...,Latest household survey,Source of most recent Income and expenditure data,Vital registration complete,Latest agricultural census,Latest industrial data,Latest trade data,Latest water withdrawal data,Unnamed: 31,Groupe de pays bis,Groupe de pays
5,ARB,Arab World,Arab World,Arab World,1A,NaN,Arab World aggregate. Arab World is composed o...,NaN,NaN,1A,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
57,EAP,East Asia & Pacific (developing only),East Asia & Pacific,East Asia & Pacific (developing only),4E,NaN,East Asia and Pacific regional aggregate (does...,NaN,NaN,4E,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
58,EAS,East Asia & Pacific (all income levels),East Asia & Pacific (all income levels),East Asia & Pacific (all income levels),Z4,NaN,East Asia and Pacific regional aggregate (incl...,NaN,NaN,Z4,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
59,ECA,Europe & Central Asia (developing only),Europe & Central Asia,Europe & Central Asia (developing only),7E,NaN,Europe and Central Asia regional aggregate (do...,NaN,NaN,7E,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
60,ECS,Europe & Central Asia (all income levels),Europe & Central Asia (all income levels),Europe & Central Asia (all income levels),Z7,NaN,Europe and Central Asia regional aggregate (in...,NaN,NaN,Z7,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
63,EMU,Euro area,Euro area,Euro area,XC,NaN,Euro area aggregate.,NaN,NaN,XC,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
68,EUU,European Union,European Union,European Union,EU,NaN,European Union aggregate.,NaN,NaN,EU,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
89,HIC,High income,High income,High income,XD,NaN,High income group aggregate. High-income econo...,NaN,NaN,XD,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
92,HPC,Heavily indebted poor countries (HIPC),Heavily indebted poor countries (HIPC),Heavily indebted poor countries (HIPC),XE,NaN,Heavily indebted poor countries aggregate.,NaN,NaN,XE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1
116,LAC,Latin America & Caribbean (developing only),Latin America & Caribbean,Latin America & Caribbean (developing only),XJ,NaN,Latin America and Caribbean regional aggregate...,NaN,NaN,XJ,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,1


Ce resultat semble encourageant pour trouver et supprimer les faux pays...

#Création d'un dataframe temporaire pour faciliter la suppression, récupération des codes pays et des noms courts des pays n'ayant pas de système de comptes nationaux défini, afin de les supprimer du dataframe StatsCountr

Malgré plusieurs filtre appliqué à savoir par les variables "Systeme de comptes Nationaux" ou "Currency Unit", il reste le pays Nauru, qui n'a pas d'observation mais qui n'est pas un groupe de pays, mais bien un pays.


In [ ]:
TemporaireSupression = StatsCountry2.loc[
    StatsCountry2["Groupe de pays bis"] == 1,
    ["Groupe de pays bis", "Country Code", "Short Name"]
]
#Création d'un dataframe temporaire pour faciliter la suppression, récupération des codes pays et des noms courts des pays n'ayant pas de système de comptes nationaux défini, afin de les supprimer du dataframe StatsCountry2

Finalement, nous n'allons pas utiliser le Dataframe *TemporaireSupression* pour supprimer, on l'utilisera sans doute plus tard. Suppression par méthode simple des groupes de pays. On conserve uniquement les observations non reconnues comme *Groupe de pays* On conserve tout de même le pays Nauru malgré qu'il soit considéré comme un groupe de pays. Le résultat obtenu est bien 215 + 1 = 216 pays soit le nombre reconnu de pays et d'économies selon le *World Bank List of Economies* pour l'année 2018.

In [ ]:
StatsCountry2 = StatsCountry2[
    (StatsCountry2["Groupe de pays"] == 0) |
    (StatsCountry2["Country Code"] == "NRU")
]


# Exercice partie 2



## Etape 1 

- Parmi les jeux de données centrés sur les indicateurs, identification de la colonne qui décrit la catégorie métier à laquelle appartient chaque indicateur.

Chaque indicateur (ex: taux d’alphabétisation, dépenses éducatives, taux d’inscription)appartient à une catégorie métier (Education OUtcomes, Education Financing, Infrastructure) il faut retrouver la colonne correspondante, après recherche c'est la colonne *Topic*, elle regroupe tous les indicateurs par grandes thématiques. 

Selon les topic on peut voir que ce dataset est orienté prinicpalement vers des sujets d'éducation

| Nom Python        | Fichier source                     | Rôle |
|------------------|------------------------------------|------|
| `CountrySeries1` | EdStatsCountry-Series.csv          | Lien pays ↔ indicateurs |
| `StatsCountry2`  | EdStatsCountry.csv                 | Informations générales sur les pays |
| `StatsData3`     | EdStatsData.csv                    | Données principales (valeurs par année) |
| `StatsFootNote4` | EdStatsFootNote.csv                | Notes méthodologiques |
| `StatsSeries5`   | EdStatsSeries.csv                  | Métadonnées des indicateurs |

In [ ]:
StatsSeries5.columns


Index(['Series Code', 'Topic', 'Indicator Name', 'Short definition',
       'Long definition', 'Periodicity', 'Base Period', 'Other notes',
       'Aggregation method', 'Limitations and exceptions', 'General comments',
       'Source', 'Statistical concept and methodology',
       'Development relevance', 'Related source links'],
      dtype='str')

In [ ]:
StatsSeries5["Topic"].value_counts()
#Pour voir les catégories existantes dans la colonne "Topic" et leur nombre d'occurrences, ce qui peut aider à comprendre la répartition des données et à identifier les sujets les plus fréquents.

Topic
Learning Outcomes                                                                               1046
Attainment                                                                                       733
Education Equality                                                                               426
Secondary                                                                                        256
Primary                                                                                          248
Population                                                                                       213
Tertiary                                                                                         158
Teachers                                                                                         137
Expenditures                                                                                      93
Engaging the Private Sector (SABER)                                                  

Ci dessus une liste d'indicateur, ils semblent encore incompréhensibles 

Il nécessite de comprendre ce que veulent dire tout ces indicateurs, afin d'en séléctioner 2 ou 3 pour continuer le processus.

Pré-sélection d'indicateurs qui semblent pertinents pour une stratégie commerciale :

-LES *SABER* L'approche systémique pour de meilleurs résultats éducatifs (SABER), lancée par la Banque mondiale en 2011, est une initiative visant à produire des données et des connaissances comparatives sur les politiques et les institutions éducatives, dans le but d'aider les pays à renforcer systématiquement leurs systèmes éducatifs et l'objectif ultime de promouvoir l'apprentissage pour tous. 

- Les dépenses : *Expenditures*
- L'égalité en éducation : *Education Equality*
- L'implication du secteur privé : *Engaging the Private Sector*
- L'enseignement supérieur : *Tertiary Education (SABER)*
- L'enseignement secondaire : *Secondary*
- La population : *Population*
- La completion des études : *Completion*


La séléction et la visualisation de ces indicateurs donnent 1056 indicateurs, chaque indicateur donnent plusieurs sous indicateurs très précis. Il nécessitera une compréhension accrue pour les utiliser convenablement.

Rappel du besoin en stratégie : StartUp qui propore des formation en ligne pour un public de niveau lycée et Université. Besoin d'explorer les pays qui ont offrent de grandes opportunité dans ces secteurs.



In [ ]:
topics = [
    "Expenditures",
    "Education Equality",
    "Engaging the Private Sector (SABER)",
    "Tertiary Education (SABER)",
    "Secondary",
    "Population",
    "Completion"
]

filtered = StatsSeries5.loc[
    StatsSeries5["Topic"].isin(topics),
    ["Topic", "Indicator Name"]
]

filtered
#Pour voir les indicateurs associés à une liste de topics spécifiques, ce qui peut aider à comprendre les types d'indicateurs liés à ces sujets et à identifier les indicateurs spécifiques qui pourraient être pertinents pour l'analyse des données sur ces sujets.


,Topic,Indicator Name
360,Education Equality,DHS: Gross attendance rate. Post Secondary
361,Education Equality,DHS: Gross attendance rate. Post Secondary. Fe...
362,Education Equality,DHS: Gross attendance rate. Post Secondary. Male
363,Education Equality,DHS: Gross attendance rate. Post Secondary. Qu...
364,Education Equality,DHS: Gross attendance rate. Post Secondary. Qu...
...,...,...
3660,Expenditures,Government expenditure per upper secondary stu...
3661,Expenditures,Government expenditure per post-secondary non-...
3662,Expenditures,Government expenditure per tertiary student (c...
3663,Expenditures,Government expenditure in secondary institutio...


Après avoir séléctionné des indicateurs pour Mark, on va supprimer tous les autres


In [ ]:
StatsSeries5 = StatsSeries5[StatsSeries5["Topic"].isin([
    "Expenditures",
    "Education Equality",
    "Engaging the Private Sector (SABER)",
    "Tertiary Education (SABER)",
    "Secondary",
    "Population",
    "Completion"
])]


Après avoir conservé cette séléction d'indicateurs, l'on va filtrer l’ensemble des jeux de données pour ne garder que les indicateurs sélectionnés.

Un indicateur fais la jonctions entre plusieurs dataset, dans notre cas c'est la variable *Serie Code* qui permet ce filtrage.

In [ ]:
#Filtrage dans le dataset 3 *StatsData3* pour ne conserver que les indicateurs liés aux topics sélectionnés dans StatsSeries5, ce qui peut aider à se concentrer sur les données pertinentes pour l'analyse des sujets d'intérêt et à réduire la quantité de données à traiter.

StatsData3 = StatsData3[
    StatsData3["Indicator Code"].isin(StatsSeries5["Series Code"])
]


In [ ]:
StatsData3["Indicator Code"].nunique()
#Le nombre d'indicateurs uniques dans StatsData3 après le filtrage


1041

In [ ]:
StatsSeries5["Series Code"].nunique()
#Le nombre d'indicateurs uniques dans StatsSeries5 après le filtrage

1056

Parmi les 1056 indicateurs sélectionnés, 1041 disposent effectivement de données observables.15 indicateurs sont définis dans la base mais ne présentent aucune valeur exploitable.

In [ ]:
StatsSeries5[StatsSeries5["Series Code"].isin(set(StatsSeries5["Series Code"]) - set(StatsData3["Indicator Code"]))][["Series Code", "Indicator Name", "Long definition"]]

#Lignes de StatsSeries5 dont les "Series Code" ne sont pas présents dans les "Indicator Code" de StatsData3, ce qui peut aider à identifier les indicateurs pour lesquels il n'y a pas de données disponibles dans StatsData3, et à comprendre les raisons possibles de cette absence de données.

,Series Code,Indicator Name,Long definition
2292,SE.SEC.DURS.LO,Theoretical duration of lower secondary educat...,Number of grades (years) in lower secondary ed...
2314,SE.SEC.ENRR.UP.FE,"Gross enrolment ratio, upper secondary, female...",Total female enrollment in upper secondary edu...
2818,UIS.E.2.Pu.F,"Enrolment in lower secondary education, public...",Total number of females enrolled in public low...
2823,UIS.E.23.Pu,"Enrolment in secondary education, public insti...",Total number of individuals enrolled in public...
2824,UIS.E.23.Pu.F,"Enrolment in secondary education, public insti...",Total number of females enrolled in public sec...
2832,UIS.E.3.Pu.F,"Enrolment in upper secondary education, public...",Total number of females enrolled in public upp...
3549,UIS.thAge.3.A.GPV,Official entrance age to upper secondary educa...,Age at which students would enter upper second...
3553,UIS.TranRA.23.GPV.GPI,Effective transition rate from primary to lowe...,The ratio of the female transition rate to the...
3582,UIS.XGDP.0.FSgov.FDinstadm.FFd,Government expenditure in pre-primary institut...,"Total general (local, regional and central) go..."
3584,UIS.XGDP.1.FSgov.FDinstadm.FFd,Government expenditure in primary institutions...,"Total general (local, regional and central) go..."


Filtrage des de la séléction des indicateurs pour les datasets restants (Le dataset 2 *StatsCountry2* contient les métadonnées des pays, donc pas de colonne indicateurs)


In [ ]:
CountrySeries1 = CountrySeries1[CountrySeries1["SeriesCode"].isin(StatsSeries5["Series Code"])]
#Filtre Data1 pour retirer les lignes dont les "SeriesCode" ne sont pas présents dans les "Series Code" de StatsSeries5, ce qui peut aider à se concentrer sur les données pertinentes pour l'analyse des sujets d'intérêt et à réduire la quantité de données à traiter.


In [ ]:
StatsData3 = StatsData3[StatsData3["Indicator Code"].isin(StatsSeries5["Series Code"])]


In [ ]:
StatsFootNote4 = StatsFootNote4[StatsFootNote4["SeriesCode"].isin(StatsSeries5["Series Code"])]


Après avoir fait un filtrage, on tente de faire une vérification de la cohérence des indicateurs. Les résultats suivants montrent qu'il reste des indicateurs sans donnée.

Attention, il se peut que des indicateurs aient des données mais pas de métadonnées

In [ ]:
set(StatsData3["Indicator Code"]) == set(StatsSeries5["Series Code"])


False

In [ ]:
StatsSeries5[
    ~StatsSeries5["Series Code"].isin(StatsData3["Indicator Code"])
][["Series Code", "Indicator Name", "Topic"]]

#Visualiser les indicateurs présents dans StatsSeries5 mais absents de StatsData3



,Series Code,Indicator Name,Topic
2292,SE.SEC.DURS.LO,Theoretical duration of lower secondary educat...,Secondary
2314,SE.SEC.ENRR.UP.FE,"Gross enrolment ratio, upper secondary, female...",Secondary
2818,UIS.E.2.Pu.F,"Enrolment in lower secondary education, public...",Secondary
2823,UIS.E.23.Pu,"Enrolment in secondary education, public insti...",Secondary
2824,UIS.E.23.Pu.F,"Enrolment in secondary education, public insti...",Secondary
2832,UIS.E.3.Pu.F,"Enrolment in upper secondary education, public...",Secondary
3549,UIS.thAge.3.A.GPV,Official entrance age to upper secondary educa...,Secondary
3553,UIS.TranRA.23.GPV.GPI,Effective transition rate from primary to lowe...,Secondary
3582,UIS.XGDP.0.FSgov.FDinstadm.FFd,Government expenditure in pre-primary institut...,Expenditures
3584,UIS.XGDP.1.FSgov.FDinstadm.FFd,Government expenditure in primary institutions...,Expenditures


In [ ]:
StatsData3[
    ~StatsData3["Indicator Code"].isin(StatsSeries5["Series Code"])
][["Indicator Code"]].drop_duplicates()
#Visualisation inverse StatsSeries5 <-> StatsData3

,Indicator Code


Maintenant on va passer à l'interprétation des années dans le dataset *StatsData3* qui contient un intervale d'année de 1970 à 2100.


NB : Il semble qu'à partir de 2017, les données (si existantes) soient prospectées par la Banque Mondiale, ce qui voudrait dire que le ficher et les observations datent de 2017.


Pour notre mission de développement commerciales, il est bon de considérer l'ensemble des années. Mais il faut considérer également des années d'intérêt, si on se place en 2017. Les années ex-ante jusqu'à N-10 maximum sont intéressantes, car précédement c'est trop ancien. A moins de réflechir à une analyse temporelle et économétrique sur le temps très long. Ensuite, les années (les observation) ex-post à 2017 peuvent être intéressante mais uniquement celles à N+5 ou N+10 grand maximum, étant entendu que les méthodologies de prospective et d'analyse de tendance de la Banque Mondiale décroissent en qualité de prédiction d'une année future sur l'autre, du fait du caractère absolument incertain de varibales socio-politico-économiques futures et incertaines.

Filtrage des années à N-10 et N-10 des autour de 2017

In [ ]:
StatsData3 = StatsData3[['Country Name','Country Code','Indicator Name','Indicator Code'] + [c for c in StatsData3.columns if any(str(y) in str(c) for y in list(range(2007,2018)) + list(range(2020,2028)))]]
